src/stm32f1/timer1_pwm.cpp

```diff
@@ -0,0 +1,203 @@
+  pwm_reg_t* const reg = get_pwm_reg(peripheral_id);
+
+  bit_modify(reg->control_register)
+    .insert<clock_division>(0b00U)
+    .insert<edge_aligned_mode>(0b00U)
+    .clear(direction);
+
+  if (p_pin == timer1_pwm::pwm_pins::pa8) {  // channel 1
+    // Preload enable must be done for corresponding OCxPE
+    bit_modify(reg->capture_compare_mode_register)
+      .insert<output_compare_odd>(pwm_mode_1)
+      .set(odd_channel_preload_enable);
+    setup_channel(reg, 1);
+  } else if (p_pin == timer1_pwm::pwm_pins::pa9) {  // channel 2
+    bit_modify(reg->capture_compare_mode_register)
+      .insert<output_compare_even>(pwm_mode_1)
+      .set(even_channel_preload_enable);
+    setup_channel(reg, 2);
+
+  } else if (p_pin == timer1_pwm::pwm_pins::pa10) {  // channel 3
+    bit_modify(reg->capture_compare_mode_register_2)
+      .insert<output_compare_odd>(pwm_mode_1)
+      .set(odd_channel_preload_enable);
+    setup_channel(reg, 3);
+
+  } else if (p_pin == timer1_pwm::pwm_pins::pa11) {  // channel 4
+    bit_modify(reg->capture_compare_mode_register_2)
+      .insert<output_compare_even>(pwm_mode_1)
+      .set(even_channel_preload_enable);
+    setup_channel(reg, 4);
+  }
+
+  bit_modify(reg->control_register).set(counter_enable);
+
+  reg->prescale_register = 0x0U;
+
+  // Set counter value to 0 because it upcounts.
+  reg->counter_register = 0x0U;
+
+  reg->auto_reload_register = 0xFFFF;
+}
+
+}  // namespace
+
+timer1_pwm::timer1_pwm(timer1_pwm::pwm_pins p_pin)
+  : m_pin{ p_pin }
+{
+  // config pin to alternate function push - pull
+  auto const p_id = get_peripheral_id(p_pin);
+  auto const p_reg = get_pwm_reg(p_id);
+  power_on(p_id);
+  // power_on(peripheral::afio);
+  switch (p_pin) {
+    case timer1_pwm::pwm_pins::pa10:
+      configure_pin({ .port = 'A', .pin = 10 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register_3;
+      m_channel = 3;
+      break;
+    case timer1_pwm::pwm_pins::pa11:
+      configure_pin({ .port = 'A', .pin = 11 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register_4;
+      m_channel = 4;
+      break;
+    case timer1_pwm::pwm_pins::pa8:
+      configure_pin({ .port = 'A', .pin = 8 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register;
+      m_channel = 1;
+      break;
+    case timer1_pwm::pwm_pins::pa9:
+      configure_pin({ .port = 'A', .pin = 9 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register_2;
+      m_channel = 2;
+      break;
+    default:
+      safe_throw(hal::operation_not_supported(this));
+  }
+  setup(p_pin);
+}
+void timer1_pwm::driver_frequency(hertz p_frequency)
+{
+  auto const peripheral_id = get_peripheral_id(m_pin);
+  pwm_reg_t* reg = get_pwm_reg(peripheral_id);
+
+  auto const current_clock = hal::stm32f1::frequency(peripheral_id);
```

Comment:
```suggestion
  auto const input_clock_frequency = hal::stm32f1::frequency(peripheral_id);
```

Current clock makes me think that this is the current PWM frequency. Which would mean that the following lines only allows the frequency to decrease but never increase.